### ISSUE TITLE
Protect API routes with withMiddlewareAuthRequired

### ISSUE THREAD

**Issue by Nargonath** [auth0/nextjs-auth0#903]
### Checklist

- [X] I have looked into the [README](https://github.com/auth0/nextjs-auth0#readme) and have not found a suitable solution or answer.
- [X] I have looked into the [examples](https://github.com/auth0/nextjs-auth0/blob/main/EXAMPLES.md) and have not found a suitable solution or answer.
- [X] I have looked into the [API documentation](https://auth0.github.io/nextjs-auth0/) and have not found a suitable solution or answer.
- [X] I have searched the [issues](https://github.com/auth0/nextjs-auth0/issues) and have not found a suitable solution or answer.
- [X] I have searched the [Auth0 Community](https://community.auth0.com/c/sdks/5) forums and have not found a suitable solution or answer.
- [X] I agree to the terms within the [Auth0 Code of Conduct](https://github.com/auth0/open-source-template/blob/master/CODE-OF-CONDUCT.md).

### Describe the problem you'd like to have solved

I have a NextJS server that has both page routes and API endpoints under `/api`. I want to use `withMiddlewareAuthRequired` to validate session token for all page routes and API endpoints that are not ignored by `config.matcher`. Once `@auth0/nextjs-auth0` did its magic, I want to execute custom code to check for the presence of specific roles and custom claims in the token. If these elements are not present, depending on the entry point (page route or API endpoint) I'd either redirect the user to Auth0 login or return a 401.

I can already craft my custom logic for the middleware by providing `withMiddlewareAuthRequired` a function. However right now when using `withMiddlewareAuthRequired` if I attempt to access a protected API route without any authentication information, the client gets redirected to the Auth0 login. 

### Describe the ideal solution

IMO redirection is the desired behaviour when accessing page routes however when you access any protected endpoint under `/api` it should return a 401 without redirecting the client.

### Alternatives and current workarounds

I don't have any since the way the redirect is handled is outside of the user control. One way would be to ignore all API routes but it defeats the purpose of middleware IMO. I'd have to craft the token validation myself and apply it manually to all protected API endpoints which is error prone.

### Additional context

next@12.3.2
@auth0/nextjs-auth0@2.0.0-beta.1
node@16.15.1

**Comment by Nargonath** [comment]
I forgot to mention that with such behaviour I could use `withMiddlewareAuthRequired` solely instead of `withPageAuthRequired` and `withApiAuthRequired` on each individual routes. I believe the middleware is more secured because you're strict by default and then you loosen it only when needed rather than being lax by default.

**Comment by adamjmcgrath** [comment]
Hi @Nargonath - thanks for raising this

This is tricky because of https://nextjs.org/docs/messages/returning-response-body-in-middleware - the advice is to [redirect to an API route](https://nextjs.org/docs/messages/returning-response-body-in-middleware#:~:text=error%2C%20login%20forms%2C-,or%20to%20an%20API%20Route.,-Before) in these cases. But that API route will also have to check the session, so there's little point using middleware for this use case - you'd just be checking the session in 2 places rather than 1.

**Comment by adamjmcgrath** [comment]
@Nargonath - have raised #909 to fix

**Comment by Nargonath** [comment]
Thanks for the explanation @adamjmcgrath. I wasn't aware we couldn't return a body from a middleware anymore. I agree in that case it doesn't make sense to use `withMiddlewareAuthRequired` for the API endpoints anymore. I'm probably better off creating a custom wrapper around `withApiAuthRequired` to also check the token claims once `withApiAuthRequired` has validated the session token, something like:

```js
import { withApiAuthRequired, getSession } from "@auth0/nextjs-auth0";

export function withCustomAuthCheck(nextHandler) {
  return withApiAuthRequired(async (req, res) => {
    const session = await getSession(req, res);

    if (!session.roles.includes("super-role")) {
      res.status(401).end();
      return;
    }

    return nextHandler(req, res);
  })
}
```

**Comment by Nargonath** [comment]
Thanks @adamjmcgrath for the explanation.

> I would say Middleware is great for protecting pages with one caveat, if the page you're protecting needs to access the session, to get user info or an access token, you'll end up doing the work of decrypting the session cookie twice - which is probably not a big deal, but worth bearing in mind.

Thanks for the heads up. Does it mean I can't do something like below?

```js
// pages/profile.js
import { getSession } from '@auth0/nextjs-auth0';

export default function Profile({ user }) {
  return <div>Hello {user.name}</div>;
}

export function getServerSideProps(ctx) {
  const { user } = await getSession(ctx.req, ctx.res);

  return { props: { user } };
}
```

**Comment by adamjmcgrath** [comment]
> Or just that the session cookie will be decrypted one time at middleware level and then one time again in the getServerSideProps?

Yep, this is correct. I don't want to put people off using `withMiddlewareAuthRequired`, but you can't share the work of decrypting the cookie between the middleware layer and the server layer so you will be decrypting it twice in that scenario you've shared.

**Comment by adamjmcgrath** [comment]
This got released in as of [`2.0.0-beta.4`](https://github.com/auth0/nextjs-auth0/releases/tag/v2.0.0-beta.4)
